server/message_dispatch.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strings"
+	"sync"
+)
+
+type MessageDispatch struct {
+	input       chan string
+	output      map[string][]chan string
```

Comment:
Not sure, why prefer iterating over array of channels instead of use simpler `Cond` structure.

Keep in mind, that sending by default is not asynchronous and if channel is not buffered (and currently, it's not), this can block dispatcher loop when some client disconnects. See example:
https://play.golang.org/p/-I9eVBimNA
https://play.golang.org/p/1yLFOtu1Cy